ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala

```diff
@@ -252,9 +254,57 @@ private[kvutils] class TransactionCommitter(
   private def blind: Step =
     (commitContext, transactionEntry) => {
       val blindingInfo = Blinding.blind(transactionEntry.transaction)
-      buildFinalResult(commitContext, transactionEntry, blindingInfo)
+      buildFinalResultState(commitContext, transactionEntry, blindingInfo)
     }
 
+  private def removeUnnecessaryNodes: Step = (commitContext, transactionEntry) => {
+    val transaction = transactionEntry.submission.getTransaction
+    val nodes: Map[String, TransactionOuterClass.Node] =
+      transaction.getNodesList.asScala.map(n => n.getNodeId -> n)(scala.collection.breakOut)
+    val nodesToKeep: Set[String] = nodes.collect {
+      case (id, node) if node.hasCreate || node.hasExercise => id
+    }(scala.collection.breakOut)
+
+    val roots = transaction.getRootsList.asScala.filter(nodesToKeep)
+
+    val stripUnnecessaryChildren: TransactionOuterClass.Node => TransactionOuterClass.Node = {
+      case node if node.hasExercise =>
+        val exerciseNode = node.getExercise
+        val keep = exerciseNode.getChildrenList.asScala.filter(nodesToKeep)
+        node.toBuilder
+          .setExercise(
+            exerciseNode.toBuilder
+              .clearChildren()
+              .addAllChildren(keep.asJavaCollection)
+              .build()
+          )
+          .build()
+      case node => node
+    }
+
+    val newRoots = roots.filter(nodesToKeep)
+
+    val newNodes = nodes
+      .collect {
+        case (id, node) if nodesToKeep(id) =>
+          stripUnnecessaryChildren(node)
+      }
+
+    val newTx = transaction
+      .newBuilderForType()
+      .addAllRoots(newRoots.asJavaCollection)
```

Comment:
Transaction version is lost here. Might be better to clear roots and nodes and add them back instead so we don't forget other fields (and new ones might be added).

Reply:
Thanks @dajmaki . I noticed, I've added it back.